File: cli/python/src/memgo_cli/cli/commands/memory/add.py
"""记忆 add 用例。"""

from __future__ import annotations

import json
import sys
from pathlib import Path

import typer

from memgo_cli.backend.json import json_object, json_messages
from rich.console import Console

from memgo_cli.application.memory.expiration import _validate_expires
from memgo_cli.backend.json import JsonObject, json_records, json_string
from memgo_cli.backend.types import Backend
from memgo_cli.output.branding import (
    print_error,
    print_scope,
    print_success,
    timed_status,
)
from memgo_cli.output.format import (
    format_add_result,
    format_agent_envelope,
)
from memgo_cli.runtime.input import _stdin_is_piped

console = Console()
err_console = Console(stderr=True)


def cmd_add(
    backend: Backend,
    text: str | None,
    *,
    user_id: str | None,
    agent_id: str | None,
    app_id: str | None,
    run_id: str | None,
    messages: str | None,
    file: Path | None,
    metadata: str | None,
    immutable: bool,
    no_infer: bool,
    expires: str | None,
    categories: str | None,
    custom_instructions: str | None,
    agent_custom_instructions: str | None,
    custom_categories: str | None,
    structured_data_schema: str | None,
    timestamp: int | None,
    output: str,
) -> None:
    """读取输入并添加记忆，按输出模式呈现处理结果。"""
    from memgo_cli.runtime.state import is_agent_mode, set_current_command

    set_current_command("add")
    if is_agent_mode():
        output = "agent"

    if categories:
        print_error(
            err_console,
            "--categories is not supported on add. Use --custom-categories instead.",
            hint=None,
        )
        raise typer.Exit(1)

    msgs = None
    content = text

    # 读取文件内容
    if file:
        try:
            raw = Path(file).read_text()
            msgs = json_messages(json.loads(raw))
        except (OSError, ValueError) as e:
            print_error(err_console, f"Failed to read file: {e}", hint=None)
            raise typer.Exit(1) from None

    # 解析消息 JSON
    elif messages:
        try:
            msgs = json_messages(json.loads(messages))
        except ValueError as e:
            print_error(err_console, f"Invalid JSON in --messages: {e}", hint=None)
            raise typer.Exit(1) from None

    # 仅在真实管道或文件重定向时读取标准输入
    elif not content and _stdin_is_piped():
        content = sys.stdin.read().strip()

    if not content and not msgs:
        print_error(
            err_console,
            "No content provided. Pass text, --messages, --file, or pipe via stdin.",
            hint=None,
        )
        raise typer.Exit(1)

    meta = None
    if metadata:
        try:
            meta = json_object(json.loads(metadata))
        except ValueError:
            print_error(err_console, "Invalid JSON in --metadata.", hint=None)
            raise typer.Exit(1) from None

    custom_cats = None
    if custom_categories:
        try:
            custom_cats = json_records(json.loads(custom_categories))
        except ValueError:
            print_error(err_console, "Invalid JSON in --custom-categories.", hint=None)
            raise typer.Exit(1) from None

    schema = None
    if structured_data_schema:
        try:
            schema = json_object(json.loads(structured_data_schema))
        except ValueError:
            print_error(err_console, "Invalid JSON in --structured-data-schema.", hint=None)
            raise typer.Exit(1) from None

    if expires:
        _validate_expires(expires)

    with timed_status(err_console, "Adding memory...") as ts:
        try:
            result = backend.add(
                content=content,
                messages=msgs,
                user_id=user_id,
                agent_id=agent_id,
                app_id=app_id,
                run_id=run_id,
                metadata=meta,
                immutable=immutable,
                infer=not no_infer,
                expires=expires,
                custom_instructions=custom_instructions,
                agent_custom_instructions=agent_custom_instructions,
                custom_categories=custom_cats,
                structured_data_schema=schema,
                timestamp=timestamp,
            )
        except Exception as e:
            ts.error_msg = str(e)
            raise typer.Exit(1) from None

    if output == "quiet":
        return

    # 所有输出模式统一按 event_id 合并 PENDING 项
    results_list = json_records(
        result if isinstance(result, list) else result.get("results", [result])
    )
    seen_events: set[str] = set()
    deduped: list[JsonObject] = []
    for r in results_list:
        if r.get("status") == "PENDING":
            eid = json_string(r.get("event_id", ""), "event_id")
            if eid and eid in seen_events:
                continue
            if eid:
                seen_events.add(eid)
        deduped.append(r)
    # 更新局部结果，供后续格式化使用
    if isinstance(result, dict) and "results" in result:
        result = {**result, "results": deduped}
    else:
        result = deduped

    if output == "agent":
        scope = {
            k: v
            for k, v in {
                "user_id": user_id,
                "agent_id": agent_id,
                "app_id": app_id,
                "run_id": run_id,
            }.items()
            if v
        }
        format_agent_envelope(
            console,
            command="add",
            data=deduped,
            scope=scope or None,
            count=len(deduped),
            duration_ms=None,
        )
        return

    if output == "json":
        format_add_result(console, result, output)
        return

    console.print()
    print_scope(console, user_id=user_id, agent_id=agent_id, app_id=app_id, run_id=run_id)
    count = len(deduped)
    all_pending = count > 0 and all(r.get("status") == "PENDING" for r in deduped)
    if all_pending:
        print_success(
            console,
            f"Memory queued — {count} event{'s' if count != 1 else ''} pending",
        )
    else:
        print_success(
            console, f"Memory processed — {count} memor{'y' if count == 1 else 'ies'} extracted"
        )
    format_add_result(console, result, output)

File: cli/node/src/cli/commands/memory/add.ts
import fs from 'node:fs';
import { _validateExpires } from '../../../application/memory/expiration.js';
import { parseCategories, parseMessages, parseObject } from '../../../application/memory/input.js';
import type { JsonObject } from '../../../backend/json.js';
import type { Backend } from '../../../backend/types.js';
import { printError, printScope, printSuccess, timedStatus } from '../../../output/branding.js';
import { formatAddResult, formatAgentEnvelope } from '../../../output/format.js';
import { exit } from '../../../runtime/exit.js';
import { setCurrentCommand, stdinIsPiped } from '../../../runtime/state.js';

/**
 * 解析文本、文件或消息输入，添加记忆并输出结果。
 * @param backend - 已装配的后端连接器。
 * @param text - 待解析或提交的文本。
 * @param opts - 本次操作选项。
 * @param opts.userId - 用户 ID。
 * @param opts.agentId - 代理 ID。
 * @param opts.appId - 应用 ID。
 * @param opts.runId - 会话 ID。
 * @param opts.messages - 消息列表的 JSON 文本。
 * @param opts.file - 记忆输入文件路径。
 * @param opts.metadata - 记忆元数据的 JSON 文本。
 * @param opts.immutable - 记忆不可变选项。
 * @param opts.infer - 是否启用记忆提取。
 * @param opts.expires - 记忆到期日期，格式为 YYYY-MM-DD。
 * @param opts.categories - 记忆分类筛选或标记。
 * @param opts.customInstructions - 记忆提取的自定义指令。
 * @param opts.agentCustomInstructions - 代理记忆的自定义指令。
 * @param opts.customCategories - 自定义分类描述的 JSON 文本。
 * @param opts.structuredDataSchema - 结构化提取模式的 JSON 文本。
 * @param opts.timestamp - 业务时间戳。
 * @param opts.output - 终端输出格式。
 */
export async function cmdAdd(
  backend: Backend,
  text: string | undefined,
  opts: {
    userId?: string;
    agentId?: string;
    appId?: string;
    runId?: string;
    messages?: string;
    file?: string;
    metadata?: string;
    immutable: boolean;
    infer?: boolean;
    expires?: string;
    categories?: string;
    customInstructions?: string;
    agentCustomInstructions?: string;
    customCategories?: string;
    structuredDataSchema?: string;
    timestamp?: number;
    output: string;
  }
): Promise<void> {
  setCurrentCommand('add');

  if (opts.categories) {
    printError('--categories is not supported on add. Use --custom-categories instead.');
    exit(1);
  }

  let msgs: JsonObject[] | undefined;
  let content = text;

  // 从文件读取输入。
  if (opts.file) {
    try {
      const raw = fs.readFileSync(opts.file, 'utf-8');
      msgs = parseMessages(raw);
    } catch (e) {
      printError(`Failed to read file: ${e instanceof Error ? e.message : e}`);
      exit(1);
    }
  }
  // 解析消息 JSON。
  else if (opts.messages) {
    try {
      msgs = parseMessages(opts.messages);
    } catch (e) {
      printError(`Invalid JSON in --messages: ${e instanceof Error ? e.message : e}`);
      exit(1);
    }
  }
  // 只从实际管道或重定向文件读取标准输入。
  else if (!content && stdinIsPiped()) {
    content = fs.readFileSync(0, 'utf-8').trim();
  }

  if (content !== undefined && content.trim() === '') {
    printError('Content cannot be empty.');
    exit(1);
  }
  if (!content && !msgs) {
    printError('No content provided. Pass text, --messages, --file, or pipe via stdin.');
    exit(1);
  }

  let meta: JsonObject | undefined;
  if (opts.metadata) {
    try {
      meta = parseObject(opts.metadata);
    } catch {
      printError('Invalid JSON in --metadata.');
      exit(1);
    }
  }

  let customCats: Record<string, string>[] | undefined;
  if (opts.customCategories) {
    try {
      customCats = parseCategories(opts.customCategories);
    } catch {
      printError('Invalid JSON in --custom-categories.');
      exit(1);
    }
  }

  let schema: JsonObject | undefined;
  if (opts.structuredDataSchema) {
    try {
      schema = parseObject(opts.structuredDataSchema);
    } catch {
      printError('Invalid JSON in --structured-data-schema.');
      exit(1);
    }
  }

  if (opts.expires) _validateExpires(opts.expires);

  let result: JsonObject | JsonObject[];
  try {
    result = await timedStatus('Adding memory...', async () => {
      return backend.add(content ?? undefined, msgs, {
        userId: opts.userId,
        agentId: opts.agentId,
        appId: opts.appId,
        runId: opts.runId,
        metadata: meta,
        immutable: opts.immutable,
        infer: opts.infer !== false,
        expires: opts.expires,
        customInstructions: opts.customInstructions,
        agentCustomInstructions: opts.agentCustomInstructions,
        customCategories: customCats,
        structuredDataSchema: schema,
        timestamp: opts.timestamp,
      });
    });
  } catch (e) {
    printError(e instanceof Error ? e.message : String(e));
    exit(1);
  }

  if (opts.output === 'quiet') return;

  // 所有输出模式按 event_id 去重待处理项。
  const rawResults: JsonObject[] = Array.isArray(result)
    ? result
    : ((result.results as JsonObject[]) ?? [result]);
  const seenEvents = new Set<string>();
  const deduped: JsonObject[] = [];
  for (const r of rawResults) {
    if (r.status === 'PENDING') {
      const eid = (r.event_id as string) ?? '';
      if (eid && seenEvents.has(eid)) continue;
      if (eid) seenEvents.add(eid);
    }
    deduped.push(r);
  }
  // 将去重结果传给输出模块。
  const dedupedResult: JsonObject | JsonObject[] = Array.isArray(result)
    ? deduped
    : { ...result, results: deduped };

  if (opts.output === 'agent') {
    const scope: Record<string, string | undefined> = {
      user_id: opts.userId,
      agent_id: opts.agentId,
      app_id: opts.appId,
      run_id: opts.runId,
    };
    formatAgentEnvelope({
      command: 'add',
      data: deduped,
      scope,
      count: deduped.length,
    });
    return;
  }

  if (opts.output === 'json') {
    formatAddResult(dedupedResult, opts.output);
    return;
  }

  console.log();
  printScope({
    user_id: opts.userId,
    agent_id: opts.agentId,
    app_id: opts.appId,
    run_id: opts.runId,
  });
  const count = deduped.length;
  const allPending = count > 0 && deduped.every((r) => r.status === 'PENDING');
  if (allPending) {
    printSuccess(`Memory queued — ${count} event${count !== 1 ? 's' : ''} pending`);
  } else {
    printSuccess(`Memory processed — ${count} memor${count === 1 ? 'y' : 'ies'} extracted`);
  }
  formatAddResult(dedupedResult, opts.output);
}
